frame/indices/src/lib.rs

```diff
@@ -82,36 +74,145 @@ decl_event!(
+		IndexFreed(AccountIndex),
 	}
 );
 
-decl_storage! {
-	trait Store for Module<T: Trait> as Indices {
-		/// The next free enumeration set.
-		pub NextEnumSet get(fn next_enum_set) build(|config: &GenesisConfig<T>| {
-			(config.ids.len() as u32 / ENUM_SET_SIZE).into()
-		}): T::AccountIndex;
-
-		/// The enumeration sets.
-		pub EnumSet get(fn enum_set) build(|config: &GenesisConfig<T>| {
-			(0..((config.ids.len() as u32) + ENUM_SET_SIZE - 1) / ENUM_SET_SIZE)
-				.map(|i| (
-					i.into(),
-					config.ids[
-						(i * ENUM_SET_SIZE) as usize..
-						config.ids.len().min(((i + 1) * ENUM_SET_SIZE) as usize)
-					].to_owned(),
-				))
-				.collect::<Vec<_>>()
-		}): map hasher(blake2_256) T::AccountIndex => Vec<T::AccountId>;
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// The index was not already assigned.
+		NotAssigned,
+		/// The index is assigned to another account.
+		NotOwner,
+		/// The index was not available.
+		InUse,
+		/// The source and destination accounts are identical.
+		NotTransfer,
 	}
-	add_extra_genesis {
-		config(ids): Vec<T::AccountId>;
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin, system = frame_system {
+		fn deposit_event() = default;
+
+		fn on_initialize() {
+			Self::migrations();
+		}
+
+		/// Assign an previously unassigned index.
+		///
+		/// Payment: `Deposit` is reserved from the sender account.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `index`: the index to be claimed. This must not be in use.
+		///
+		/// Emits `IndexAssigned` if successful.
+		///
+		/// # <weight>
+		/// - `O(1)`.
+		/// - One storage mutation (codec `O(1)`).
+		/// - One reserve operation.
+		/// - One event.
+		/// # </weight>
+		fn claim(origin, index: T::AccountIndex) {
+			let who = ensure_signed(origin)?;
+
+			Accounts::<T>::try_mutate(index, |maybe_value| {
+				ensure!(maybe_value.is_none(), Error::<T>::InUse);
+				*maybe_value = Some((who.clone(), T::Deposit::get()));
+				T::Currency::reserve(&who, T::Deposit::get())
+			})?;
+			Self::deposit_event(RawEvent::IndexAssigned(who, index));
+		}
+
+		/// Assign an index already owned by the sender to another account. The balance reservation
+		/// is effectively transfered to the new account.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `index`: the index to be re-assigned. This must be owned by the sender.
+		/// - `new`: the new owner of the index. This function is a no-op if it is equal to sender.
+		///
+		/// Emits `IndexAssigned` if successful.
+		///
+		/// # <weight>
+		/// - `O(1)`.
+		/// - One storage mutation (codec `O(1)`).
+		/// - One transfer operation.
+		/// - One event.
+		/// # </weight>
+		fn transfer(origin, new: T::AccountId, index: T::AccountIndex) {
+			let who = ensure_signed(origin)?;
+			ensure!(who != new, Error::<T>::NotTransfer);
+
+			Accounts::<T>::try_mutate(index, |maybe_value| {
```

Comment:
How does `try_mutate` work in terms of the lower storage layers?

For example, your first line in here is:

```
let (account, amount) = maybe_value.take().ok_or(Error::<T>::NotAssigned)?;
```

This is one storage read, and if this fails, we need not do any more logic. If `try_mutate` also invokes a storage read before any of this logic is run I would say that this method of writing the function possibly invokes an additional storage read in situations when it would not be necessary.